src/App.tsx

```diff
@@ -1,14 +1,69 @@
 /* eslint-disable max-len */
-import React from 'react';
+import React, { useEffect, useState } from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { Filter } from './types/Filter';
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
+
+const getFilteredTodos = (
+  todos: Todo[],
+  filter: Filter,
+  query: string,
+): Todo[] => {
+  const normalizedQuery = query.trim().toLowerCase();
+
+  let filteredTodos = todos.filter(todo => {
+    if (filter === Filter.Completed) {
+      return todo.completed;
+    }
+
+    if (filter === Filter.Active) {
+      return !todo.completed;
+    }
+
+    return true;
+  });
+
+  filteredTodos = filteredTodos.filter(todo => {
+    const normalizedTodoTitle = todo.title.trim().toLowerCase();
+
+    return normalizedTodoTitle.includes(normalizedQuery);
+  });
+
+  return filteredTodos;
+};
 
 export const App: React.FC = () => {
+  const [todosFromServer, setTodosFromServer] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(false);
+  const [query, setQuery] = useState('');
+  const [filter, setFilter] = useState<Filter>(Filter.All);
+  const [selectedTodoId, setSelectedTodoId] = useState(0);
+
+  useEffect(() => {
+    setIsLoading(true);
+    getTodos().then(todos => {
+      setTodosFromServer(todos);
+      setIsLoading(false);
+    });
+  }, []);
+
+  const handleOptionFilter = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    const value = event.target.value as Filter;
+
+    setFilter(value);
+  };
+
+  const filteredTodos = getFilteredTodos(todosFromServer, filter, query);
```

Comment:
The `filteredTodos` are recalculated on each render. Consider using `useMemo` hook to avoid unnecessary recalculations. The `useMemo` will only recompute the memoized value when one of the dependencies has changed.